shinydashboard/lantern/global.R

```diff
@@ -47,93 +49,38 @@ app <<- list(
-  well_known_endpoint_counts = NULL,  # summary table of well known URI endpoints
-  endpoint_locations = NULL,          # endpoints with location information mappings
-  implementation_guide = NULL         # implementation_guide table
+  fhir_version_list = reactiveVal(NULL),           # list of fhir_versions reported by endpoints
+  fhir_endpoint_totals = reactiveVal(NULL),        # count of endpoints, indexed and nonindexed
+  response_tally = reactiveVal(NULL),              # counts of http responses
+  http_pct = reactiveVal(NULL),                    # percentage of http responses for each endpoint
+  http_summary = reactiveVal(NULL),                # counts of all http_responses ever
+  vendor_count_tbl = reactiveVal(NULL),            # endpoint counts by vendor
+  endpoint_resource_types = reactiveVal(NULL),     # Resource types from capability statement by endpoint
+  capstat_fields = reactiveVal(NULL),              # fields from the capability statement
+  capstat_fields_list = reactiveVal(NULL),         # the list of fields we keep track of in a capability statement
+  capstat_values = reactiveVal(NULL),              # values of specific fields from the capability statement
+  last_updated = reactiveVal(NULL),                # time app_data was last updated
+  avg_response_time = reactiveVal(NULL),           # mean response time for endpoints by refresh period
+  vc_totals = reactiveVal(NULL),                   # counts of endpoints by vendor
+  security_endpoints = reactiveVal(NULL),          # security auth types supported by each endpoint
+  security_endpoints_tbl = reactiveVal(NULL),      # list of endpoints filterable by auth type
+  auth_type_counts = reactiveVal(NULL),            # count and pct of endpoints by auth type and fhir_version
+  endpoint_security_counts = reactiveVal(NULL),    # summary table of endpoint counts with security resource in cap statement
+  security_code_list = reactiveVal(NULL),          # list of supported auth types for UI dropdown
+  smart_response_capabilities = reactiveVal(NULL), # smart core capabilities by endpoint, vendor, fhir_version
+  well_known_endpoints_tbl = reactiveVal(NULL),    # endpoints returning smart core capabilities JSON doc
+  well_known_endpoints_no_doc = reactiveVal(NULL), # well known endpoints reached, but no JSON doc returned
+  well_known_endpoint_counts = reactiveVal(NULL),  # summary table of well known URI endpoints
+  endpoint_locations = reactiveVal(NULL),          # endpoints with location information mappings
+  implementation_guide = reactiveVal(NULL)         # implementation_guide table
 )
 
 # Define observer based on a refresh_timeout to refetch data from the database
 updater <- observe({
 
   invalidateLater(config_yaml$refresh_timeout_minutes * 60 * 1000) # convert minutes to milliseconds
+  database_fetch(1)
 
-  app$fhir_version_list <<- get_fhir_version_list(endpoint_export_tbl)
```

Comment:
From my understanding, all of these variables were added to the global.R file so they could not only run on initial startup and refreshed as a specific interval, but also so that they could be used globally from any file. Therefore, the initial declaration of these app$data tables need to occur in the global.R file, but then since I want them to be populated while a spinner is displayed on the UI, they must be populated within a file, such as the endpoints.R file, that is called after the initial UI startup, as all the global.R functions happen before everything else and therefore a spinner would not be able to be displayed from this file. However, since we need to declare the app$data tables in global.R (so that they can be refreshed at a specific interval and also remain globally accessible), they needed to be converted to reactiveVals since only then could their values be reset from NULL to the successfully loaded tables within the endpoints.R file. Therefore, moving everything to endpoints.R changes the functionality in that if I did not have them declared in the global.R file first, they would no longer be global variables, so they must be declared in the global.R file. However, in order to show a spinner, they must be loaded in a function that is called after the UI is set up, meaning they have to be reactiveVals whose values can be updated within the endpoints.R file

Reply:
ahaaaa okay, thanks for the thorough clarification, that was really useful!